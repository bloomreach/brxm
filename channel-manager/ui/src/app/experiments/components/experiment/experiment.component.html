<!--
  Copyright 2020 Bloomreach. All rights reserved. (https://www.bloomreach.com/)

  Licensed under the Apache License, Version 2.0 (the "License");
  you may not use this file except in compliance with the License.
  You may obtain a copy of the License at

   http://www.apache.org/licenses/LICENSE-2.0

  Unless required by applicable law or agreed to in writing, software
  distributed under the License is distributed on an "AS IS" BASIS,
  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  See the License for the specific language governing permissions and
  limitations under the License.
  -->

<ng-container *ngIf="experiment$ | async as experiment; else experimentForm">
  <em-experiment-info [experiment]="experiment"></em-experiment-info>

  <div class="experiment-status-wrapper">
    <em-experiment-status-message [winnerVariant]="experiment.winnerVariant"
                                  [visits]="experiment.totalVisits"
                                  class="qa-experiment-status-message">
    </em-experiment-status-message>
    <em-experiment-status-chart [experiment]="experiment"></em-experiment-status-chart>
  </div>
</ng-container>

<ng-template #experimentForm>
  <em-experiment-start-form [variants]="availableVariants$ | async"
                            [goals]="availableGoals$ | async">
  </em-experiment-start-form>
</ng-template>
