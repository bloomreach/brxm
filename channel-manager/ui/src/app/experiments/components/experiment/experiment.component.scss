/*
 * Copyright 2020-2023 Bloomreach
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *  http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

@import '~@bloomreach/frontend-theme/sizes';

.experiment {
  height: 100%;
  display: grid;
  grid-template-rows: auto 1fr auto;
  grid-template-areas:
    "info"
    "status"
    "actions";
}

em-experiment-info {
  grid-area: info;
}

.experiment-status-wrapper {
  grid-area: status;
  padding: $br-size;
}

.actions-wrapper {
  grid-area: actions;
}

.actions {
  padding: $br-size-m $br-size;
  text-align: right;
}

.variants-info {
  list-style-type: '- ';
  padding: 0;

  > li {
    padding-top: $br-size-s;
  }
}
