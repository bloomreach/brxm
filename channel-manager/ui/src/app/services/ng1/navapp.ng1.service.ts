/*!
 * Copyright 2021 Bloomreach. All rights reserved. (https://www.bloomreach.com/)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *  http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { InjectionToken } from '@angular/core';
import { ClientError, NavLocation, ParentConfig, UserSettings } from '@bloomreach/navapp-communication';

export interface Ng1NavappService {
  subscribe: (api: string, callback: () => Promise<any>) => void;
  unsubscribe: (api: string, callback: () => Promise<any>) => void;
  connect: () => Promise<void>;
  getConfig(): Promise<ParentConfig>;
  getUserSettings(): Promise<UserSettings>;
  updateNavLocation(location: NavLocation): Promise<void>;
  navigate(location: NavLocation): Promise<void>;
  showMask(): Promise<void>;
  hideMask(): Promise<void>;
  showBusyIndicator(): Promise<void>;
  hideBusyIndicator(): Promise<void>;
  onUserActivity(): Promise<void>;
  onSessionExpired(): Promise<void>;
  onError(error: ClientError): Promise<void>;
}

export const NG1_NAVAPP_SERVICE = new InjectionToken<Ng1NavappService>('NG1_NAVAPP_SERVICE');
