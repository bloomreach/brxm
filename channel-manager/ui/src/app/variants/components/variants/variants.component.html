<!--
  Copyright 2020 Bloomreach. All rights reserved. (https://www.bloomreach.com/)

  Licensed under the Apache License, Version 2.0 (the "License");
  you may not use this file except in compliance with the License.
  You may obtain a copy of the License at

   http://www.apache.org/licenses/LICENSE-2.0

  Unless required by applicable law or agreed to in writing, software
  distributed under the License is distributed on an "AS IS" BASIS,
  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  See the License for the specific language governing permissions and
  limitations under the License.
  -->

<div class="mat-typography">
  <div class="variant-select">
    <mat-form-field appearance="outline">
      <mat-label class="qa-variant-select-label">
        {{ 'VARIANT' | translate }}

      </mat-label>
      <mat-select [formControl]="variantSelect"
                  class="qa-variant-select"
                  (selectionChange)="selectVariant($event.value)">
        <mat-option *ngFor="let variant of variants"
                    class="{{ 'qa-variant-select-option-' + variant.id }}"
                    [value]="variant">
          {{ variant.variantName }}
        </mat-option>
      </mat-select>
    </mat-form-field>

    <button mat-icon-button
            class="qa-variant-copy"
            (click)="addVariant()">
      <mat-icon svgIcon="copy"></mat-icon>
    </button>

    <button mat-icon-button
            class="qa-variant-delete">
      <mat-icon svgIcon="trash"></mat-icon>
    </button>
  </div>

  <div class="variant-expressions"
       *ngIf="currentVariant?.expressions.length > 0">
    <div class="variant-expressions-list">
      <div class="variant-expressions-list-label">
        Variant Expressions
      </div>

      <div class="variant-expressions-list-item"
           *ngFor="let expression of currentVariant?.expressions">
        <span>{{ expression.name }}</span>
        <button mat-icon-button (click)="removeExpression(expression)">
          <mat-icon svgIcon="close"></mat-icon>
        </button>
      </div>
    </div>

    <div class="variant-expressions-actions">
      <button mat-stroked-button>
        <mat-icon svgIcon="plus"></mat-icon>
        Add segment
      </button>
    </div>
  </div>
</div>
