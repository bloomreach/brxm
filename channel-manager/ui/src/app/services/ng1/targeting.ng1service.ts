/*!
 * Copyright 2020 Bloomreach. All rights reserved. (https://www.bloomreach.com/)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *  http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { InjectionToken } from '@angular/core';

import { ExperimentGoal } from '../../experiments/models/experiment-goal.model';
import { ExperimentStatus } from '../../experiments/models/experiment-status.model';
import { Experiment } from '../../experiments/models/experiment.model';
import { TargetingApiResponse } from '../../models/targeting-api-response.model';
import { Variant, VariantCharacteristicData } from '../../variants/models/variant.model';

export interface Ng1TargetingService {
  getCharacteristics(): Promise<TargetingApiResponse<void>>;
  getVariants(containerItemId: string): Promise<TargetingApiResponse<Variant[]>>;
  addVariant(
    componentId: string,
    formData: any,
    personaId?: string,
    characteristics?: VariantCharacteristicData[],
  ): Promise<TargetingApiResponse<any>>;
  getGoals(): Promise<TargetingApiResponse<ExperimentGoal[]>>;
  getExperiment(componentId: string): Promise<TargetingApiResponse<Experiment>>;
  getExperimentStatus(experimentId: string): Promise<TargetingApiResponse<ExperimentStatus>>;
  saveExperiment(componentId: string, goalId: string, variantId: string): Promise<TargetingApiResponse<string>>;
  completeExperiment(componentId: string, keepOnlyVariantId: string): Promise<TargetingApiResponse<void>>;
}

export const NG1_TARGETING_SERVICE = new InjectionToken<Ng1TargetingService>('NG1_TARGETING_SERVICE');
