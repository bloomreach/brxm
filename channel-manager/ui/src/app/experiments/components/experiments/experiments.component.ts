/*!
 * Copyright 2020 Bloomreach. All rights reserved. (https://www.bloomreach.com/)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *  http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { Component, Inject } from '@angular/core';

import { Ng1ComponentEditorService, NG1_COMPONENT_EDITOR_SERVICE } from '../../../services/ng1/component-editor.ng1.service';
import { ExperimentsService } from '../../services/experiments.service';

@Component({
  templateUrl: 'experiments.component.html',
  styleUrls: ['experiments.component.scss'],
})
export class ExperimentsComponent {
  private readonly component = this.componentEditorService.getComponent();
  private readonly componentId = this.component.getId();
  readonly experiment$ = this.experimentsService.getExperiment(this.componentId);

  constructor(
    @Inject(NG1_COMPONENT_EDITOR_SERVICE) private readonly componentEditorService: Ng1ComponentEditorService,
    private readonly experimentsService: ExperimentsService,
  ) {}
}
