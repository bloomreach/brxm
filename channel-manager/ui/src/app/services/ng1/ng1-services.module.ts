/*!
 * Copyright 2020 Bloomreach. All rights reserved. (https://www.bloomreach.com/)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *  http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { NgModule } from '@angular/core';

import { NG1_CHANNEL_SERVICE } from './channel.ng1service';
import { NG1_CONFIG_SERVICE } from './config.ng1.service';
import { NG1_CONTENT_SERVICE } from './content.ng1.service';
import { NG1_IFRAME_SERVICE } from './iframe.ng1service';
import { NG1_PAGE_STRUCTURE_SERVICE } from './page-structure.ng1.service';
import { NG1_PAGE_SERVICE } from './page.ng1.service';
import { NG1_PROJECT_SERVICE } from './project.ng1.service';
import { NG1_ROOT_SCOPE } from './root-scope.service';
import { NG1_WORKFLOW_SERVICE } from './workflow.ng1.service';

@NgModule({
  providers: [
    { provide: NG1_CHANNEL_SERVICE, useValue: window.angular.element(document.body).injector().get('ChannelService') },
    { provide: NG1_CONFIG_SERVICE, useValue: window.angular.element(document.body).injector().get('ConfigService') },
    { provide: NG1_CONTENT_SERVICE, useValue: window.angular.element(document.body).injector().get('ContentService') },
    { provide: NG1_IFRAME_SERVICE, useValue: window.angular.element(document.body).injector().get('HippoIframeService') },
    { provide: NG1_PAGE_STRUCTURE_SERVICE, useValue: window.angular.element(document.body).injector().get('PageStructureService') },
    { provide: NG1_PAGE_SERVICE, useValue: window.angular.element(document.body).injector().get('PageService') },
    { provide: NG1_PROJECT_SERVICE, useValue: window.angular.element(document.body).injector().get('ProjectService') },
    { provide: NG1_WORKFLOW_SERVICE, useValue: window.angular.element(document.body).injector().get('WorkflowService') },
    { provide: NG1_ROOT_SCOPE, useValue: window.angular.element(document.body).injector().get('$rootScope') },
  ],
})
export class Ng1ServicesModule {
}
