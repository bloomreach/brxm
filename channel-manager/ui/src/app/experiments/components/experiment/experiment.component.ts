/*!
 * Copyright 2020 Bloomreach. All rights reserved. (https://www.bloomreach.com/)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *  http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { Component, Inject } from '@angular/core';
import { TranslateService } from '@ngx-translate/core';

import { Ng1ComponentEditorService, NG1_COMPONENT_EDITOR_SERVICE } from '../../../services/ng1/component-editor.ng1.service';
import { NotificationService } from '../../../services/notification.service';
import { VariantsService } from '../../../variants/services/variants.service';
import { ExperimentState } from '../../models/experiment-state.enum';
import { Experiment } from '../../models/experiment.model';
import { ExperimentsService } from '../../services/experiments.service';
import { SelectedVariantIdAndGoalId } from '../experiment-start-form/experiment-start-form.component';

@Component({
  templateUrl: 'experiment.component.html',
  styleUrls: ['experiment.component.scss'],
})
export class ExperimentComponent {
  private readonly component = this.componentEditorService.getComponent();
  private readonly componentId = this.component.getId();
  readonly availableVariants$ = this.variantsService.getVariants(this.componentId);
  readonly availableGoals$ = this.experimentsService.getGoals();

  experiment$ = this.experimentsService.getExperiment(this.componentId);

  requestInProgress = false;

  constructor(
    @Inject(NG1_COMPONENT_EDITOR_SERVICE) private readonly componentEditorService: Ng1ComponentEditorService,
    private readonly experimentsService: ExperimentsService,
    private readonly variantsService: VariantsService,
    private readonly notificationService: NotificationService,
    private readonly translateService: TranslateService,
  ) {}

  isExperimentRunning(experiment: Experiment): boolean {
    return experiment.state === ExperimentState.Running;
  }

  isExperimentRunningOrCompleted(experiment: Experiment): boolean {
    return this.isExperimentRunning(experiment) || experiment.state === ExperimentState.Completed;
  }

  async onVariantAndGoalSelected(value: SelectedVariantIdAndGoalId): Promise<void> {
    const { variantId, goalId } = value;

    try {
      this.requestInProgress = true;

      await this.experimentsService.saveExperiment(this.componentId, variantId, goalId);

      this.experiment$ = this.experimentsService.getExperiment(this.componentId);

      this.notificationService.showNotification(this.translateService.instant('EXPERIMENT_SAVED'));
    } catch (e) {
      this.notificationService.showErrorNotification(this.translateService.instant('EXPERIMENT_SAVE_ERROR'));
    } finally {
      this.requestInProgress = false;
    }
  }

  async onCompleteExperiment(experiment: Experiment): Promise<void> {
    const firstNonDefaultVariant = experiment.variants.find(v => v.variantId !== this.variantsService.defaultVariantId);

    if (!firstNonDefaultVariant) {
      throw new Error('Unable to find a non default variant');
    }

    try {
      this.requestInProgress = true;

      await this.experimentsService.completeExperiment(this.componentId, firstNonDefaultVariant.variantId);

      this.experiment$ = this.experimentsService.getExperiment(this.componentId);

      this.notificationService.showNotification(this.translateService.instant('EXPERIMENT_COMPLETED'));
    } catch {
      this.notificationService.showErrorNotification(this.translateService.instant('EXPERIMENT_COMPLETION_ERROR'));
    } finally {
      this.requestInProgress = false;
    }
  }
}
