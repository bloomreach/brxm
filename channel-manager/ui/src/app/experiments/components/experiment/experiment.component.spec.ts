/*!
 * Copyright 2020-2021 Bloomreach. All rights reserved. (https://www.bloomreach.com/)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *  http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { NO_ERRORS_SCHEMA, Pipe, PipeTransform } from '@angular/core';
import { ComponentFixture, fakeAsync, TestBed, tick } from '@angular/core/testing';
import { TranslateService } from '@ngx-translate/core';
import { mocked } from 'ts-jest/utils';

import { NG1_CHANNEL_SERVICE } from '../../../services/ng1/channel.ng1service';
import { NG1_COMPONENT_EDITOR_SERVICE } from '../../../services/ng1/component-editor.ng1.service';
import { NotificationService } from '../../../services/notification.service';
import { VariantsService } from '../../../variants/services/variants.service';
import { ExperimentState } from '../../models/experiment-state.enum';
import { ExperimentWithStatusData } from '../../models/experiment-with-status-data.model';
import { ExperimentsService } from '../../services/experiments.service';

import { ExperimentComponent } from './experiment.component';

@Pipe({name: 'translate'})
export class TranslateMockPipe implements PipeTransform {
  transform(value: string): string {
    return value;
  }
}

describe('ExperimentComponent', () => {
  let experimentsService: ExperimentsService;
  let notificationService: NotificationService;

  const mockExperiment: ExperimentWithStatusData = {
    id: 'experiment-1',
    state: ExperimentState.Running,
    type: 'PAGE',
    startTime: 1600868196063,
    winnerVariant: null,
    goal: {
      id: 'goal-1',
      name: 'Goal 1',
      type: 'PAGE',
      readOnly: false,
      targetPage: '/target-page',
      mountId: 'some-mount-id',
    },
    variants: [
      {
        variantId: 'default',
        variantName: 'Default',
        confidence: 0.3,
        mean: 0.2,
        variance: 0.4,
        visitorSegment: 'segment-1',
      },
      {
        variantId: 'some-variant',
        variantName: 'Some variant',
        confidence: 0.3,
        mean: 0.2,
        variance: 0.4,
        visitorSegment: 'segment-1',
      },
    ],
    statusWithVisits: [
      { 'variant-1': 3, default: 1, timestamp: 1600948800000, visits: 4 },
      { 'variant-1': 100, default: 10, timestamp: 1600959600000, visits: 110 },
      { 'variant-1': 40, default: 1000, timestamp: 1600970400000, visits: 1040 },
      { 'variant-1': 20, default: 5, timestamp: 1600981200000, visits: 25 },
      { 'variant-1': 80, default: 80, timestamp: 1600992000000, visits: 160 },
      { 'variant-1': 11, default: 12, timestamp: 1601002800000, visits: 23 },
      { 'variant-1': 0, default: 0, timestamp: 1601013600000, visits: 0 },
      { 'variant-1': 3, default: 4, timestamp: 1601024400000, visits: 7 },
      { 'variant-1': 1000, default: 500, timestamp: 1601035200000, visits: 1500 },
      { 'variant-1': 123, default: 456, timestamp: 1601046000000, visits: 579 },
      { 'variant-1': 234, default: 546, timestamp: 1601056800000, visits: 779 },
      { 'variant-1': 4635, default: 4365, timestamp: 1601067600000, visits: 9000 },
      { 'variant-1': 345, default: 465, timestamp: 1601078400000, visits: 810 },
      { 'variant-1': 465, default: 467, timestamp: 1601089200000, visits: 922 },
    ],
  };

  const mockMountId = 'some-mount-id';

  const mockGoals = [
    {
      id: 'goal-1',
      name: 'Goal 1',
      type: 'PAGE',
      readOnly: false,
      targetPage: '/target-page',
      mountId: mockMountId,
    },
    {
      id: 'goal-2',
      name: 'Goal 2',
      type: 'PAGE',
      readOnly: false,
      targetPage: '/target-page',
      mountId: mockMountId,
    },
    {
      id: 'goal-3',
      name: 'Goal 3',
      type: 'PAGE',
      readOnly: false,
      targetPage: '/some-target-page',
      mountId: 'some-other-mount-id',
    },
  ];

  const mockVariants = [
    {
      id: 'variant-1',
      variantName: 'Variant 1',
    },
    {
      id: 'variant-2',
      variantName: 'Variant 2',
    },
  ];

  const mockComponent = {
    getId: () => 'mockComponentId',
    isXPageComponent: () => false,
  };

  beforeEach(() => {
    const componentEditorServiceMock = {
      getComponent: () => mockComponent,
    };

    const experimentsServiceMock = {
      getExperiment: jest.fn(() => Promise.resolve(mockExperiment)),
      getGoals: jest.fn().mockResolvedValue(mockGoals),
      saveExperiment: jest.fn(),
      completeExperiment: jest.fn(),
    };

    const variantsServiceMock = {
      defaultVariantId: 'hippo-default',
      getVariants: jest.fn().mockResolvedValue(mockVariants),
    };

    const notificationServiceMock = {
      showNotification: jest.fn(),
      showErrorNotification: jest.fn(),
    };

    const translateServiceMock = {
      instant: jest.fn(x => x),
    };

    const channelServiceMock = {
      getChannel: jest.fn(() => ({ mountId: mockMountId })),
    };

    TestBed.configureTestingModule({
      declarations: [
        ExperimentComponent,
        TranslateMockPipe,
      ],
      providers: [
        { provide: NG1_CHANNEL_SERVICE, useValue: channelServiceMock },
        { provide: NG1_COMPONENT_EDITOR_SERVICE, useValue: componentEditorServiceMock },
        { provide: ExperimentsService, useValue: experimentsServiceMock },
        { provide: VariantsService, useValue: variantsServiceMock },
        { provide: NotificationService, useValue: notificationServiceMock },
        { provide: TranslateService, useValue: translateServiceMock },
      ],
      schemas: [
        NO_ERRORS_SCHEMA,
      ],
    });

    experimentsService = TestBed.inject(ExperimentsService);
    notificationService = TestBed.inject(NotificationService);
  });

  describe('if experiment was not saved', () => {
    let fixture: ComponentFixture<ExperimentComponent>;
    let component: ExperimentComponent;

    beforeEach(fakeAsync(() => {
      fixture = TestBed.createComponent(ExperimentComponent);

      fixture.detectChanges();
      tick();
      fixture.detectChanges();

      component = fixture.componentInstance;
    }));

    it('should show the experiment form', fakeAsync(() => {
      expect(fixture.nativeElement).toMatchSnapshot();
    }));

    it('should only show goals from current channel', async () => {
      const filteredGoals = mockGoals.filter(g => g.mountId === mockMountId);
      const availableGoals = await component.availableGoals$;
      expect(availableGoals).toEqual(filteredGoals);
    });

    describe('onCancelExperiment', () => {
      it('should cancel the experiment', () => {
        component.onCancelExperiment();

        expect(experimentsService.saveExperiment).toHaveBeenCalledWith('mockComponentId');
      });

      it('should load the saved experiment', async () => {
        const mockSavedExperiment = { ...mockExperiment , state: ExperimentState.Created };

        mocked(experimentsService.saveExperiment).mockResolvedValue();
        mocked(experimentsService.getExperiment).mockResolvedValue(mockSavedExperiment);

        await component.onCancelExperiment();
        const savedExperiment = component.experiment;

        expect(experimentsService.getExperiment).toHaveBeenCalledWith('mockComponentId');
        expect(savedExperiment).toBe(mockSavedExperiment);
      });

      it('should show a notification after successful saving', async () => {
        mocked(experimentsService.saveExperiment).mockResolvedValue();

        await component.onCancelExperiment();

        expect(notificationService.showNotification).toHaveBeenCalledWith('EXPERIMENT_CANCELED');
      });

      it('should show an error notification after unsuccessful saving', async () => {
        mocked(experimentsService.saveExperiment).mockRejectedValue(new Error());

        await component.onCancelExperiment();

        expect(notificationService.showErrorNotification).toHaveBeenCalledWith('EXPERIMENT_CANCEL_ERROR');
      });
    });

    describe('onVariantAndGoalSelected', () => {
      const variantAndGoal = {
        variantId: 'variant-1',
        goalId: 'goal-1',
      };

      it('should save the experiment', () => {
        component.onVariantAndGoalSelected(variantAndGoal);

        expect(experimentsService.saveExperiment).toHaveBeenCalledWith('mockComponentId', 'variant-1', 'goal-1');
      });

      it('should load the saved experiment', async () => {
        const mockSavedExperiment = { ...mockExperiment , state: ExperimentState.Created };

        mocked(experimentsService.saveExperiment).mockResolvedValue();
        mocked(experimentsService.getExperiment).mockResolvedValue(mockSavedExperiment);

        await component.onVariantAndGoalSelected(variantAndGoal);
        const savedExperiment = component.experiment;

        expect(experimentsService.getExperiment).toHaveBeenCalledWith('mockComponentId');
        expect(savedExperiment).toBe(mockSavedExperiment);
      });

      it('should show a notification after successful saving', async () => {
        mocked(experimentsService.saveExperiment).mockResolvedValue();

        await component.onVariantAndGoalSelected(variantAndGoal);

        expect(notificationService.showNotification).toHaveBeenCalledWith('EXPERIMENT_SAVED');
      });

      it('should show an error notification after unsuccessful saving', async () => {
        mocked(experimentsService.saveExperiment).mockRejectedValue(new Error());

        await component.onVariantAndGoalSelected(variantAndGoal);

        expect(notificationService.showErrorNotification).toHaveBeenCalledWith('EXPERIMENT_SAVE_ERROR');
      });
    });

    describe('onCompleteExperiment', () => {
      it('should complete the experiment', () => {
        component.onCompleteExperiment('variant-1');

        expect(experimentsService.completeExperiment).toHaveBeenCalledWith('mockComponentId', 'variant-1');
      });

      it('should load the completed experiment', async () => {
        const mockSavedExperiment = { ...mockExperiment , state: ExperimentState.Created };

        mocked(experimentsService.completeExperiment).mockResolvedValue();
        mocked(experimentsService.getExperiment).mockResolvedValue(mockSavedExperiment);

        await component.onCompleteExperiment();
        const savedExperiment = component.experiment;

        expect(experimentsService.getExperiment).toHaveBeenCalledWith('mockComponentId');
        expect(savedExperiment).toBe(mockSavedExperiment);
      });

      it('should show a notification after successful completion', async () => {
        mocked(experimentsService.completeExperiment).mockResolvedValue();

        await component.onCompleteExperiment();

        expect(notificationService.showNotification).toHaveBeenCalledWith('EXPERIMENT_COMPLETED');
      });

      it('should show an error notification after unsuccessful completion', async () => {
        mocked(experimentsService.completeExperiment).mockRejectedValue(new Error());

        await component.onCompleteExperiment();

        expect(notificationService.showErrorNotification).toHaveBeenCalledWith('EXPERIMENT_COMPLETION_ERROR');
      });
    });

    describe('requestInProgress', () => {
      const variantAndGoal = {
        variantId: 'variant-1',
        goalId: 'goal-1',
      };

      it('should be set to true', () => {
        component.onVariantAndGoalSelected(variantAndGoal);

        expect(component.requestInProgress).toBeTruthy();
      });

      it('should be set to false on successful experiment saving', async () => {
        await component.onVariantAndGoalSelected(variantAndGoal);

        expect(component.requestInProgress).toBeFalsy();
      });

      it('should be set to false on unsuccessful experiment saving', async () => {
        mocked(experimentsService.saveExperiment).mockRejectedValue(new Error());

        await component.onVariantAndGoalSelected(variantAndGoal);

        expect(component.requestInProgress).toBeFalsy();
      });
    });
  });

  describe('if experiment was saved', () => {
    let fixture: ComponentFixture<ExperimentComponent>;

    beforeEach(fakeAsync(() => {
      fixture = TestBed.createComponent(ExperimentComponent);

      fixture.detectChanges();
      tick();
      fixture.detectChanges();
    }));

    it('should show the experiment', fakeAsync(() => {
      expect(fixture.nativeElement).toMatchSnapshot();
    }));
  });
});
