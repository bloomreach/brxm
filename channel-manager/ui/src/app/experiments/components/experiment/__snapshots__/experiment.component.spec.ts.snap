// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ExperimentComponent should be shown 1`] = `
<div
  id="root0"
>
  
  
  <mat-list
    role="list"
  >
    <mat-list-item
      role="listitem"
    >
      <span
        class="experiment-goal-header"
      >
        100
      </span>
      <mat-divider />
    </mat-list-item>
    <mat-list-item
      role="listitem"
    >
      <span
        class="line-header"
      >
        EXPERIMENT_GOAL
      </span>
      <span>
        Goal 1
      </span>
      <mat-divider />
    </mat-list-item>
    <mat-list-item
      role="listitem"
    >
      <span
        class="line-header"
      >
        EXPERIMENT_PAGE
      </span>
      <span>
        /target-page
      </span>
      <mat-divider />
    </mat-list-item>
    <mat-list-item
      role="listitem"
    >
      <span
        class="line-header"
      >
        EXPERIMENT_STATE
      </span>
      <span
        class="experiment-state"
      >
        running
      </span>
      <mat-divider />
    </mat-list-item>
    <mat-list-item
      role="listitem"
    >
      <span
        class="line-header"
      >
        EXPERIMENT_START_DATE
      </span>
      <span>
        1600868196063
      </span>
      <mat-divider />
    </mat-list-item>
  </mat-list>
  <em-experiment-status-chart />
</div>
`;
